File: zad_2.go
package main

import (
 "fmt"
 "math/rand"
 "strconv"
 "time"
)

func swapDigits(n int) int {
 if n < 10 {
  return n
 }
 s := strconv.Itoa(n)
 runes := []rune(s)
 runes[0], runes[len(runes)-1] = runes[len(runes)-1], runes[0]
 res, _ := strconv.Atoi(string(runes))
 return res
}

func main() {
 rand.Seed(time.Now().UnixNano())

 var n1, n2 int
 fmt.Print("Введите размер первого массива (n1 >= 10): ")
 fmt.Scan(&n1)
 if n1 < 10 {
  fmt.Println("Ошибка: n1 должно быть >= 10")
  return
 }

 fmt.Print("Введите размер второго массива (n2 >= 50): ")
 fmt.Scan(&n2)
 if n2 < 50 {
  fmt.Println("Ошибка: n2 должно быть >= 50")
  return
 }

 var arr1 [1000]float64
 var arr2 [1000]int

 // 1. Заполнение первого массива
 for i := 0; i < n1; i++ {
  arr1[i] = rand.Float64() * 100
 }
 fmt.Println("\nПервый массив (вещественные числа):")
 for i := 0; i < n1; i++ {
  fmt.Printf("%.2f ", arr1[i])
 }
 fmt.Println()

 // 2. Среднее и сумма квадратов отклонений
 sum := 0.0
 for i := 0; i < n1; i++ {
  sum += arr1[i]
 }
 avg := sum / float64(n1)

 sqSum := 0.0
 for i := 0; i < n1; i++ {
  diff := arr1[i] - avg
  sqSum += diff * diff
 }
 fmt.Printf("\nСреднее значение: %.4f\n", avg)
 fmt.Printf("Сумма квадратов разностей: %.4f\n", sqSum)

 // 3. Замена разрядов
 for i := 0; i < n1; i++ {
  val := int(arr1[i])
  if i%2 == 0 && val%2 == 0 {
   arr1[i] = float64(swapDigits(val))
  }
 }
 fmt.Println("\nПосле замены разрядов:")
 for i := 0; i < n1; i++ {
  fmt.Printf("%.2f ", arr1[i])
 }
 fmt.Println()

 // 4. Перемещение первого элемента в конец
 first := arr1[0]
 for i := 0; i < n1-1; i++ {
  arr1[i] = arr1[i+1]
 }
 arr1[n1-1] = first

 fmt.Println("\nПосле модификации (первый в конец):")
 for i := 0; i < n1; i++ {
  fmt.Printf("%.2f ", arr1[i])
 }
 fmt.Println()

 // 5. Заполнение второго массива
 for i := 0; i < n2; i++ {
  arr2[i] = rand.Intn(21) - 10 // [-10;10]
 }
 fmt.Println("\nВторой массив (целые числа):")
 for i := 0; i < n2; i++ {
  fmt.Printf("%d ", arr2[i])
 }
 fmt.Println()

 // Частоты
 var counts [21]int
 for i := 0; i < n2; i++ {
  counts[arr2[i]+10]++
 }

 fmt.Println("\nЧастоты элементов:")
 for i := 0; i < 21; i++ {
  if counts[i] > 0 {
   fmt.Printf("%d: %d\n", i-10, counts[i])
  }
 }

 // Удаление повторов
 var seen [21]bool
 var unique [21]int
 uniqueCount := 0
 for i := 0; i < n2; i++ {
  idx := arr2[i] + 10
  if !seen[idx] {
   seen[idx] = true
   unique[uniqueCount] = arr2[i]
   uniqueCount++
  }
 }
 fmt.Println("\nУникальные элементы:")
 for i := 0; i < uniqueCount; i++ {
  fmt.Printf("%d ", unique[i])
 }
 fmt.Println()
}

File: zad_2.cpp
#include <iostream>
#include <random>
#include <iomanip>
#include <cmath>
#include <string>
using namespace std;

// Меняет первый и последний разряд в числе (если больше 1 цифры)
int swapDigits(int number) {
    if (number < 10) return number;
    string s = to_string(number);
    swap(s[0], s[s.length() - 1]);
    return stoi(s);
}

int main() {
    int n1, n2;
    cout << "Введите размер первого массива (n1 >= 10): ";
    cin >> n1;
    if (n1 < 10) {
        cerr << "Ошибка: n1 должно быть >= 10\n";
        return 1;
    }

    cout << "Введите размер второго массива (n2 >= 50): ";
    cin >> n2;
    if (n2 < 50) {
        cerr << "Ошибка: n2 должно быть >= 50\n";
        return 1;
    }

    double arr1[1000];
    int arr2[1000];

    // Инициализация генератора ranlux24
    random_device rd;
    ranlux24 gen(rd());
    uniform_real_distribution<> real_dist(0.0, 100.0);
    uniform_int_distribution<> int_dist(-10, 10);

    // 1. Заполнение первого массива
    for (int i = 0; i < n1; ++i)
        arr1[i] = real_dist(gen);

    cout << "\nПервый массив (вещественные числа):\n";
    for (int i = 0; i < n1; ++i)
        cout << fixed << setprecision(2) << arr1[i] << " ";
    cout << "\n";

    // 2. Среднее и сумма квадратов отклонений
    double sum = 0;
    for (int i = 0; i < n1; ++i)
        sum += arr1[i];
    double avg = sum / n1;

    double sq_sum = 0;
    for (int i = 0; i < n1; ++i)
        sq_sum += pow(arr1[i] - avg, 2);

    cout << "\nСреднее значение: " << avg << "\n";
    cout << "Сумма квадратов разностей: " << sq_sum << "\n";

    // 3. Замена разрядов в четных элементах с четным индексом
    for (int i = 0; i < n1; ++i) {
        int val = (int)arr1[i];
        if (i % 2 == 0 && val % 2 == 0)
            arr1[i] = swapDigits(val);
    }

    cout << "\nПосле замены разрядов:\n";
    for (int i = 0; i < n1; ++i)
        cout << arr1[i] << " ";
    cout << "\n";

    // 4. Перемещение первого элемента в конец
    double first = arr1[0];
    for (int i = 0; i < n1 - 1; ++i)
        arr1[i] = arr1[i + 1];
    arr1[n1 - 1] = first;

    cout << "\nПосле модификации (первый в конец):\n";
    for (int i = 0; i < n1; ++i)
        cout << arr1[i] << " ";
    cout << "\n";

    // 5. Заполнение второго массива
    for (int i = 0; i < n2; ++i)
        arr2[i] = int_dist(gen);

    cout << "\nВторой массив (целые числа):\n";
    for (int i = 0; i < n2; ++i)
        cout << arr2[i] << " ";
    cout << "\n";

    // Подсчёт повторов
    int counts[21] = {0}; // Индексы 0..20 для чисел -10..10
    for (int i = 0; i < n2; ++i)
        counts[arr2[i] + 10]++;

    cout << "\nЧастоты элементов:\n";
    for (int i = 0; i <= 20; ++i)
        if (counts[i] > 0)
            cout << (i - 10) << ": " << counts[i] << "\n";

    // Удаление повторов
    int uniqueArr[21];
    int uniqueCount = 0;
    bool seen[21] = {false};

    for (int i = 0; i < n2; ++i) {
        int idx = arr2[i] + 10;
        if (!seen[idx]) {
            seen[idx] = true;
            uniqueArr[uniqueCount++] = arr2[i];
        }
    }

    cout << "\nМассив без повторов:\n";
    for (int i = 0; i < uniqueCount; ++i)
        cout << uniqueArr[i] << " ";
    cout << "\n";

    return 0;
}
